### DISCUSSION TITLE
How to register route for a model?

### DISCUSSION THREAD

**Discussion by dunrix** [jeremyevans/autoforme#38] [General]
I've tried to follow *autoforme* description for basic configuration, but it seems route for a model is not registered:

```
module MyProj
  class App < Roda
    plugin :autoforme

    AutoForme.for(:roda, self) do
      model Product do
        columns [:name, :price]
      end
    end
  end

  DB = Sequel.connect(ENV['DATABASE_URL'] || 'sqlite:/')
  Model = Class.new(Sequel::Model)
  Model.db = DB
  DB.create_table?(:products) do
    primary_key :id
    String :name, null: false
    BigDecimal :price, size: [10,2]
    Fixnum :stock, null: false, default: 0
  end

  class Product < Model
  end
end
 ```

However when I attempt to browse products on */Product/browse* path as suggested, I get *404-error*. I've been expecting method `AutoForme.for` would also register routes to model views automatically but it does not in my case. Did I miss something in docs ? How are routes registered for models in this plugin ?

Thank you in an advance.

**Comment by jeremyevans** [comment]
Looks like there wasn't even basic instructions for setting up AutoForme with Roda.  I apologize for that.  I've updated the README to add some instructions for use with Roda in c7d4a262135e9910b8b77c9d81f3490899bd1e1d, hopefully that helps.  If you think additional information is still needed, please let me know.

**Reply by dunrix** [reply]
Thanks for the prompt reply, however still out of luck, getting 404-error. Either in case of implicit root path or explicit after `autoforme(name)` setting.

```
 class App < Roda
    plugin :autoforme do
      model Artist
    end

    route do
      # rest of routing tree
      autoforme    # causes 404 response
      # autoforme.inspect    # this results in `false`
    end
  end
```

So far figured out [`RodaPlugins::AutoForme::InstanceMethods#autoforme`](https://github.com/jeremyevans/autoforme/blob/c7d4a262135e9910b8b77c9d81f3490899bd1e1d/lib/roda/plugins/autoforme.rb#L46) creates [lambda instance](https://github.com/jeremyevans/autoforme/blob/c7d4a262135e9910b8b77c9d81f3490899bd1e1d/lib/autoforme/frameworks/roda.rb#L89) which however results in 404 response instead of autoforme view.

Inspection of `current_matchers` in route proc from `AutoForme::Frameworks::Roda#initialize` reveals `[:model, :action_type, #<Proc:0x00005639fd6fa698 /…/.bundle/ruby/3.0.0/gems/autoforme-1.9.1/lib/autoforme/frameworks/roda.rb:98 (lambda)>]` content.

Still no clue where exactly this fails.

**Reply by dunrix** [reply]
To be a bit more precise, the `@route_proc` does not match `/Artist/browse` request.

**Reply by jeremyevans** [reply]
I tested the following as a bare minimum self contained config.ru file and it seems to work for the expected paths (e.g. `/Artist/browse`):

```ruby
require 'sequel'
require 'roda'

DB = Sequel.sqlite

DB.create_table :artists do
  primary_key :id
  String :name
end

class Artist < Sequel::Model
end

class App < Roda
    plugin :render, :layout_opts=>{:inline=>'<%= yield %>'}
    plugin :sessions, :secret=>'1'*64
    plugin :flash
    plugin :autoforme do
      model Artist
    end

    route do
      autoforme
    end
end

run App.freeze.app
```

Can you post a similar bare minimum config.ru file that isn't working, and hopefully we can determine why.

**Reply by dunrix** [reply]
Thank You again.

Your example works, but *stops working when wrapped inside a module* like in my case:

``` ruby
require 'sequel'
require 'roda'

module MyProject
# copy-pasted your code sample
end

run MyProject::App.freeze.app
```
Application starts without warning/errors (puma server), then accessing the same url at `/Artist/browse` fails with 404 error.

**Reply by dunrix** [reply]
Hello, me again.

Just tried to access sequel model by *fully-qualified name* and it begun to work. In case of above example, access autoforme with `http://localhost:3000/MyProject::Artist/browse`.
Is this an intended behaviour ?

**Reply by jeremyevans** [reply]
Yes, this is intended.  Use the `link_name` configuration setting if you would like to use a different string in the link.

**Reply by dunrix** [reply]
Great, thanks for your assistance. Still in learning phase of Roda ;-)

**Reply by jeremyevans** [reply]
You are welcome.  Thanks for bringing this up!  I updated the documentation in the README for this to make it clear: 99173dd2cb395c2acd3cc77e32ccb5b92b8116df
